src/org/ods/component/CopyImageStage.groovy

```diff
@@ -0,0 +1,106 @@
+        Map<String, Object> config,
+        OpenShiftService openShift,
+        ILogger logger) {
+        super(script, context, logger)
+
+        if (config.verifyTLS == null) {
+            config.verifyTLS = true
+        }
+
+        this.options = new CopyImageOptions(config)
+        this.openShift = openShift
+        this.logger = logger
+    }
+
+    // This is called from Stage#execute if the branch being built is eligible.
+    @SuppressWarnings(['AbcMetric'])
+    @TypeChecked(TypeCheckingMode.SKIP)
+    protected run() {
+        logger.info("Copy the image ${options.sourceImageUrlIncludingRegistry}!")
+        final String STR_DOCKER_PROTOCOL = 'docker://'
+
+        logger.info("Resolved source Image data: " +
+            "${this.options.sourceImageUrlIncludingRegistry},\n" +
+            "importing into: " +
+            "${STR_DOCKER_PROTOCOL}${context.clusterRegistryAddress}/${context.cdProject}/${this.options.image}")
+
+        // to make this work, we need a target image stream first
+        openShift.findOrCreateImageStream("${context.cdProject}", "${this.options.image.split(':').first()}")
+
+        // read the target registry auth (for this project's -cd project ..)
+        def targetInternalRegistryToken = steps.readFile '/run/secrets/kubernetes.io/serviceaccount/token'
+
+        def sourcetoken = options.sourceCredential ? "--src-creds ${options.sourceCredential}" : ''
+
+        int status = copyImage(sourcetoken, targetInternalRegistryToken, STR_DOCKER_PROTOCOL)
+        if (status != 0) {
+            script.error("Could not copy `${this.options.sourceImageUrlIncludingRegistry}', status ${status}")
+        }
+
+        // Extract image name and tag for use in compliance later
+        def imageName = this.options.image.split(':').first()
+        def imageTag = this.options.image.split(':').last()
+
+        def internalImageRef = openShift.getImageReference(context.cdProject, imageName, imageTag)
+
+        logger.info("!!! Image successfully imported into " +
+            "${context.clusterRegistryAddress}/${context.cdProject}/${this.options.image},\n" +
+            "details: ${internalImageRef}")
+
+        // Add the imported image to the artifacts so they can show up in compliance docs
+        def buildResourceName = "imported-${imageName}"
+        context.addBuildToArtifactURIs("${buildResourceName}",
+            [image: internalImageRef, source: options.sourceImageUrlIncludingRegistry])
+
+        // If this option is set will additionally tag the image in the target namespace, not just the cd namespace
+        if (options.tagIntoTargetEnv) {
+            openShift.findOrCreateImageStream(context.targetProject, imageName)
+            openShift.importImageTagFromProject(
+                context.targetProject, imageName, context.cdProject,
+                imageTag, imageTag
+            )
+            logger.info("!!! Image ${this.options.image} tagged into ${context.targetProject}")
```

Comment:
Note to self and @jafarre-bi @jorge-romero for the planned activity of improving log messages, and error messages in particular, we may want to use class name prefixes instead of `!!!`. Also, any errors which are known and can be handled, we need to introduce error numbers which resolve to help documentation.